C++/Python API Parity: add AlphaDropout and FeatureAlphaDropout

suyashb95 opened the pull request:
- add `AlphaDropoutImpl` to `modules/dropout.h` and `modules/dropout.cpp`
 - add `functional/dropout.h` containing the `alpha_dropout` function
 - include `functional/dropout.h` in `nn/functional.h`
 - add functional and module tests
-  related issue https://github.com/pytorch/pytorch/issues/25883

suyashb95 committed:
C++/Python API Parity: add AlphaDropout
 - add `AlphaDropoutImpl` to `modules/dropout.h` and `modules/dropout.cpp`
 - add `functional/dropout.h` containing the `alpha_dropout` function
 - include `functional/dropout.h` in `nn/functional.h`
 - update `DropoutOptions` to have params `p` and `inplace` instead of `rate`- update `DropoutImpl` and `FeatureDropoutImpl` to use the new `DropoutOptions`
 - add `AlphaDropoutOptions` for use in `AlphaDropoutImpl`
 - update `torch::alpha_dropout` call to `F::alpha_dropout` in `modules/dropout.cpp`
 - add checks in `functional/dropout.cpp` to check the dropout rate
 - add functional/module/prettyprint tests and update existing `Dropout` tests to work with the new `DropoutOptions`

yf225 commented on torch/csrc/api/include/torch/nn/modules/dropout.h:
```diff
@@ -64,6 +64,19 @@ class TORCH_API FeatureDropoutImpl
+class TORCH_API AlphaDropoutImpl
+    : public detail::DropoutImplBase<AlphaDropoutImpl> {
+ public:
+  explicit AlphaDropoutImpl(const DropoutOptions& options_ = DropoutOptions());
```
@Suyash458 Thanks so much for the awesome work! I think we can change the arguments in `DropoutOptions` to include `p` and `inplace` instead of `rate` to match the Python version better. And we can have `using AlphaDropoutOptions = DropoutOptions;` and then use `AlphaDropoutOptions` everywhere in the AlphaDropout code :D

Suyash458 replied:
@yf225 Thanks for having a look! Adding `p` and `inplace` makes sense, I'll add the extra params to `DropoutOptions`. This'll affect `DropoutImpl` and `FeatureDropoutImpl` as well

yf225 replied:
Yes and I think we should rename `rate` to `p` in `DropoutOptions`, and it is okay to break backward compatibility in favor of code simplicity and parity with Python API

yf225 commented on torch/csrc/api/include/torch/nn/functional/dropout.h:
```diff
@@ -0,0 +1,14 @@
+namespace nn {
+namespace functional {
+
+inline Tensor alpha_dropout(const Tensor& input, const DropoutOptions& options, bool is_training) {
```
I believe we can write
```suggestion
inline Tensor alpha_dropout(const Tensor& input, const AlphaDropoutOptions& options = {}, bool training = false) {
```
to match the Python names better and also more consistent with the naming of other functional options :D

yf225 commented on torch/csrc/api/include/torch/nn/functional/dropout.h:
```diff
@@ -0,0 +1,14 @@
+namespace functional {
+
+inline Tensor alpha_dropout(const Tensor& input, const DropoutOptions& options, bool is_training) {
+  return torch::alpha_dropout(input, options.rate(), is_training);
```
It would be awesome to express the same logic as in Python version:
```python
    if p < 0. or p > 1.:
        raise ValueError("dropout probability has to be between 0 and 1, "
                         "but got {}".format(p))
    return (_VF.alpha_dropout_(input, p, training)
            if inplace
            else _VF.alpha_dropout(input, p, training))
```

Suyash458 replied:
Just wondering, wouldn't this bit of code take care of the check for the `p` value?

```c++
DropoutImplBase<Derived>::DropoutImplBase(const DropoutOptions& options_)
    : options(options_) {
  TORCH_CHECK(options.rate() >= 0, "Dropout rate must not be less than zero");
  TORCH_CHECK(options.rate() <= 1, "Dropout rate must not be greater than one");
}```

yf225 replied:
Yes I think it would take care of the check when the user is calling `F::alpha_dropout` from the `AlphaDropout` module, but the user might also call `F::alpha_dropout` directly, and they would be out of luck :D

Suyash458 replied:
Right, I've added the same checks in the `F::alpha_dropout` function

yf225 commented on torch/csrc/api/src/nn/modules/dropout.cpp:
```diff
@@ -45,5 +46,16 @@ Tensor FeatureDropoutImpl::forward(const Tensor& input) {
+    : DropoutImplBase(options_) {}
+
+Tensor AlphaDropoutImpl::forward(const Tensor& input) {
+  return torch::alpha_dropout(input, options.rate(), this->is_training());
```
I believe this can be
```suggestion
  return F::alpha_dropout(input, options.p(), this->is_training());
```
to match the Python version even better :D

yf225 commented on torch/csrc/api/src/nn/modules/dropout.cpp:
```diff
@@ -45,5 +46,16 @@ Tensor FeatureDropoutImpl::forward(const Tensor& input) {
+}
+
+void AlphaDropoutImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::AlphaDropout(rate=" << options.rate() << ")";
```
We might want to print both `p` and `inplace`, to match the Python version even better :D

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -1267,3 +1267,15 @@ TEST_F(FunctionalTest, Pad) {
+    auto output = F::alpha_dropout(input, rate, false);
+    ASSERT_TRUE(torch::allclose(input_mean, output.mean(), 0.1));
+    ASSERT_TRUE(torch::allclose(input_std, output.std(), 0.1));
+  }
```
Thanks so much for the awesome tests! It would be great if we could also add a test for the `F::alpha_dropout(input)` (i.e. no value passed for either `options` or `training`) use case :D

Suyash458 replied:
Ah right, I missed that. Will add a test case for it

yf225 replied:
Thanks a lot for your help!

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -2079,3 +2079,11 @@ TEST_F(ModulesTest, PrettyPrintThreshold) {
+  ASSERT_EQ(c10::str(AlphaDropout(
+      DropoutOptions(0.2))),
+    "torch::nn::AlphaDropout(rate=0.2)");
+}
```
It would be awesome to add a module test similar to
https://github.com/pytorch/pytorch/blob/e31adeb4f38b41de5c0f840b6cdf05593711d4da/test/cpp/api/modules.cpp#L897-L911

Suyash458 replied:
I have added a test similar to the one for `Dropout`. From what I read in [this](https://arxiv.org/pdf/1706.02515.pdf) paper, the sum of the output should lie between 40 and 130

>New Dropout Technique. Standard dropout randomly sets an activation x to zero with probability 1 − q for 0 < q 6 1 .  In order to preserve the mean, the activations are scaled by 1 /q during training.  If x has mean E(x) = μ and variance Var(x) = ν , and the dropout variable d follows a binomial distribution B(1 ,q ) , then the mean E(1 /qdx ) = μ is kept.  Dropout fits well to rectified linear units, since zero is in the low variance region and corresponds to the default value. For scaled exponential linear units, the default and low variance value is lim x →−∞ selu(x)  = − λα = α ′ . Therefore, we propose “alpha dropout”, that randomly sets inputs to α ′ .  The new mean and new variance is E(xd + α ′(1 − d)) = qμ + (1 − q)α ′ , and Var(xd + α ′(1 − d)) = q((1 − q)(α ′ − μ) 2 + ν) .  We aim at keeping mean and variance to their original values after “alpha dropout”, in order to ensure the self-normalizing property even for “alpha dropout”. The affine transformation a (xd + α ′ (1 − d)) + b allows to determine parameters a and b such that mean and variance are kept to their values: E( a ( xd + α ′ (1 − d )) + b ) = μ and Var( a ( xd + α ′ (1 − d )) + b ) = ν . In contrast to dropout, a and b will depend on μ and ν , however our SNNs converge to activations with zero mean and unit variance.  With μ = 0 and ν = 1 , we obtain a = (q + α′2q(1 − q )) − 1 / 2 and b = − ( q + α ′ 2 q (1 − q ) ) − 1 / 2 ((1 − q ) α ′ ) . The parameters a and b only depend on the dropout rate 1 − q and the most negative activation α

yf225 replied:
Thanks so much for adding the test! :D

yf225 commented on torch/csrc/api/include/torch/nn/functional.h:
```diff
@@ -9,3 +9,4 @@
 #include <torch/nn/functional/pixelshuffle.h>
 #include <torch/nn/functional/pooling.h>
 #include <torch/nn/functional/vision.h>
+#include <torch/nn/functional/dropout.h>
```
nit: would be awesome to put this line above `#include <torch/nn/functional/embedding.h>`, following alphabetical order

yf225 reviewed (commented):
@Suyash458 Thanks so much for the awesome work! I left some comments.

Suyash458 commented:
@yf225 I have added another commit with the required changes, lemme know if they look good!

yf225 commented on torch/csrc/api/include/torch/nn/functional/dropout.h:
```diff
@@ -0,0 +1,19 @@
+
+inline Tensor alpha_dropout(const Tensor& input, const AlphaDropoutOptions& options = {}, bool training = false) {
+  TORCH_CHECK(options.p() >= 0, "Dropout rate must not be less than zero");
+  TORCH_CHECK(options.p() <= 1, "Dropout rate must not be greater than one");
```
To strictly match the Python version, maybe we could do the following instead:
```cpp
TORCH_CHECK(
  options.p() >= 0 && options.p() <= 1,
  "dropout probability has to be between 0 and 1, but got ", options.p());
```

yf225 commented on torch/csrc/api/include/torch/nn/functional/dropout.h:
```diff
@@ -0,0 +1,19 @@
+  TORCH_CHECK(options.p() >= 0, "Dropout rate must not be less than zero");
+  TORCH_CHECK(options.p() <= 1, "Dropout rate must not be greater than one");
+  
+  return torch::alpha_dropout(input, options.p(), training);
```
In the Python version, I think we execute different code paths based on the value of `options.inplace()`:
```python
    return (_VF.alpha_dropout_(input, p, training)
            if inplace
            else _VF.alpha_dropout(input, p, training))
```
It would be awesome to do the same in C++ version as well (we might need to change the type of `input` from `const Tensor&` to `Tensor` in order to achieve this) :D

Suyash458 replied:
In `Functions.h` I see two declarations for `alpha_dropout`
``` cpp
static inline Tensor & alpha_dropout_(Tensor & self, double p, bool train) {...}
static inline Tensor alpha_dropout(const Tensor & self, double p, bool train) {...}
```
Guessing we have to use the first one for in-place operation?

yf225 commented on torch/csrc/api/include/torch/nn/options/dropout.h:
```diff
@@ -4,16 +4,18 @@
 #include <torch/csrc/WindowsTorchApiMacro.h>
 #include <torch/types.h>
 
+
```
nit: we can remove the new line here

yf225 commented on torch/csrc/api/include/torch/nn/options/dropout.h:
```diff
@@ -4,16 +4,18 @@
 /// Options for `Dropout` and `FeatureDropout`.
 struct TORCH_API DropoutOptions {
-  /* implicit */ DropoutOptions(double rate = 0.5);
+  /* implicit */ DropoutOptions(double p = 0.5, bool inplace = false);
```
I think we can put the default value for `inplace` as part of the `TORCH_ARG` declaration instead, something like
```cpp
TORCH_ARG(bool, inplace) = false;
```
instead of specifying it as default value to the constructor argument. (Ideally we should do the same for `p` as well, but there are probably too many people using `Dropout(0.5)` directly, and we have to preserve it for better backward compatibility.)

yf225 commented on torch/csrc/api/src/nn/modules/dropout.cpp:
```diff
@@ -8,42 +8,62 @@
-  TORCH_CHECK(options.rate() >= 0, "Dropout rate must not be less than zero");
-  TORCH_CHECK(options.rate() <= 1, "Dropout rate must not be greater than one");
+  TORCH_CHECK(options.p() >= 0, "Dropout rate must not be less than zero");
+  TORCH_CHECK(options.p() <= 1, "Dropout rate must not be greater than one");
```
To strictly match the Python version, maybe we could do the following instead:
```cpp
TORCH_CHECK(
  options.p() >= 0 && options.p() <= 1,
  "dropout probability has to be between 0 and 1, but got ", options.p());
```

yf225 reviewed (commented):
@Suyash458 Thanks so much for the update! I left some minor comments.

yf225 commented:
We will probably need to rebase the PR because of the conflicts :(

Suyash458 commented:
Rebasing to resolve conflicts in test files

Suyash458 commented:
@yf225 is the current build failing on Linux? I get an error saying 
```
../caffe2/opt/onnxifi_op.cc: In function ‘void caffe2::{anonymous}::copyDescriptor(const caffe2::ExternalTensorDescriptor*, onnxTensorDescriptorV1*)’:
../caffe2/opt/onnxifi_op.cc:95:7: error: ‘onnxTensorDescriptorV1 {aka struct onnxTensorDescriptorV1}’ has no member named ‘isOffline’
   to->isOffline = from->isOffline;
```
[Here's](https://pastebin.com/c75LN3CV) the complete stack trace

suyashb95 committed:
Merge branch 'master' of https://github.com/pytorch/pytorch

suyashb95 committed:
Merge branch 'master' of https://github.com/pytorch/pytorch

suyashb95 committed:
Merge branch 'master' of https://github.com/pytorch/pytorch

suyashb95 committed:
Merge branch 'master' of https://github.com/pytorch/pytorch

suyashb95 committed:
fix test files

suyashb95 committed:
- update `TORCH_CHECK` error messages in `modules/dropout.cpp` and `functional/dropout.h`
- update input params for `forward()` from `const Tensor&` to `Tensor` in `DropoutImpl`
- update `functional/dropout.h` to call different implementations based on the `inplace` param
- move the `inplace` default value to the `TORCH_ARG` declaration

Suyash458 commented:
@yf225 lemme know if this looks fine, I'll rebase to fix the conflicts then

Commit:
merge

Commit:
update

Commit:
update

Commit:
clean up options

Commit:
update

Commit:
update

Commit:
fix

Commit:
fix test

Commit:
fix test

Commit:
fix test

yf225 reviewed (approved):
@Suyash458 Thanks a lot for the awesome work!

Suyash458 commented:
> @Suyash458 Thanks a lot for the awesome work!

Glad I could help!
